2017/ru/0xa4-xxe.md

```diff
@@ -0,0 +1,71 @@
+* хотя бы один из XML-обработчиков приложения или веб-службы на основе SOAP использует [определение типа документов (DTD)](https://en.wikipedia.org/wiki/Document_type_definition). Поскольку механизм отключения DTD зависит от обработчика, рекомендуется воспользоваться справочной информацией, например, ["Памяткой OWASP по предотвращению XXE"](https://www.owasp.org/index.php/XML_External_Entity_(XXE)_Prevention_Cheat_Sheet); 
+* приложение использует SAML для идентификации в рамках федеративной безопасности или технологии единого входа (SSO). SAML использует XML для подтверждения идентификаторов, поэтому может быть уязвим;
+* приложение использует SOAP версии ниже 1.2. Оно может быть уязвимо для XXE-атак, если XML-сущности передаются фреймворку SOAP;
+* если приложение уязвимо для XXE-атак, то злоумышленник может также вызвать отказ в обслуживании или осуществить атаку с использованием миллиона XML-сущностей (Billion Laughs).
+
+## Как предотвратить
+
+Обучение разработчиков имеет большое значение для выявления и противодействия XXE. Кроме того, для предотвращения XXE необходимо:
+
+* использовать, по возможности, более простые форматы данных, например, JSON, и избегать сериализации критичных данных;
+* установить исправления или обновления для всех библиотек и обработчиков XML, используемых приложением или ОС. Использовать проверки зависимостей. Обновить SOAP до версии 1.2 или выше;
+* отключить обработку внешних сущностей XML и DTD во всех XML-обработчиках приложения, согласно ["Памятке OWASP по предотвращению XXE"](https://www.owasp.org/index.php/XML_External_Entity_(XXE)_Prevention_Cheat_Sheet); 
+* реализовать на сервере (по белым спискам) проверку, фильтрацию или очистку (экранирование) входных данных для предотвращения попадания вредоносных данных в XML-документы, заголовки или узлы;
+* удостовериться, что функция загрузки XML или XSL проверяет входящие файлы с использованием XSD или другой подобной методики;
+* анализировать код масштабных и сложных приложений со множеством встраиваемых компонентов вручную, хотя инструменты SAST могут помочь обнаружить XXE в исходном коде.
+
+Если выполнение данных требований не возможно, попробуйте использовать виртуальные патчи, шлюзы безопасности API или файрволы веб-приложений (WAF) для обнаружения, мониторинга и блокировки XXE-атак.
+
+## Примеры сценариев атак
+
+Было зафиксировано большое количество XXE-атак, включая атаки на встроенные устройства. XXE обнаруживаются в самых неожиданных местах, включая глубоко вложенные зависимости. Самым простым способом реализации атаки является загрузка (если поддерживается) вредоносного XML-файла:
+
+**Сценарий №1**: Злоумышленник пытается получить данные с сервера:
+
+```
+  <?xml version="1.0" encoding="ISO-8859-1"?>
+    <!DOCTYPE foo [
+    <!ELEMENT foo ANY >
+    <!ENTITY xxe SYSTEM "file:///etc/passwd" >]>
+    <foo>&xxe;</foo>
+```
+
+**Сценарий №2**: Злоумышленник исследует внутреннюю сеть сервера, заменяя вышеуказанную строку ENTITY на:
+```
```

Comment:
Always use line break before the code snippet.

Reply:
OK. Done here.